enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,613 @@
+  red --> reboot
+```
+
+### API Extensions
+
+None
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Since this is a must-have functionality, the risks are not foreseeing fail scenarios in advance and implementation bugs.
+
+To mitigate the risks, a thorough review of the enhancement must be done by MicroShift, OpenShift, and RHEL teams,
+and making sure testing strategy is sound and prioritized equally with the feature development.
+
+### Drawbacks
+
+N/A
+
+## Design Details
+
+### Definitions
+
+- **ostree commit**: TODO
+- **ostree deployment**: TODO
+- **Rollback**: booting previous ostree commit due to greenboot assessing system to not be functional
+- **Upgrade**: running newer version of MicroShift than previously as a result of booting another ostree commit
+- **Downgrade**: running older version of MicroShift than previously as a result of booting another ostree commit
+- **Backup**: backing up `/var/lib/microshift`
+- **Restore**: restoring `/var/lib/microshift`
+- **Version Metadata**: File residing in MicroShift data dir containing version of MicroShift that successfully started (cluster was healthy)
+- **MicroShift greenboot healthcheck**: Program verifying the status of MicroShift's cluster
+
+### Preface
+
+Every action related to procedure described in this enhancement is 
+performed after system's boot rather than immediately before shutdown. 
+Greenboot's healthchecks, green and red scripts are executed independent of MicroShift's processes.
+Actions related to backup, restore, and data migration are performed with MicroShift 
+partially running, i.e. only etcd and kube-apiserver are running.
+
+Only one backup of MicroShift data will be stored at a given moment
+due to high probability of devices having limited storage.
+
+### System rollback and failed data restore detection
+
+TODO: How to not fall into endless loop of rollbacks 
+
+### Integration with greenboot
+
+Depending on result of greenboot's healthcheck either "green" (successful boot) or "red" (unsuccessful) scripts are executed before rebooting the system.
+
+MicroShift will integrate with that system to persist an action to perform on next boot:
+- "green": on next boot, before MicroShift starts, make a backup of MicroShift's data
+- "red": on next boot, before MicroShift starts, restore MicroShift's data from backup
+
+Functionality will be implemented by placing in `/etc/greenboot/green.d` and `/etc/greenboot/red.d`
+bash scripts containing with simple logic or, if needed,
+executing commands `microshift greenboot green` and `microshift greenboot red` in case of needing to put more information into file with action for next boot.
+Alternatively, a symlinks to `microshift` binary can be made and MicroShift modified to run specific command depending on content of argv[0] (just like BusyBox).
+
+File containing said information should not be part of MicroShift data directory
+as we don't to be a part of backup.
```

Comment:
```suggestion
as we don't want it to be a part of backup.
```

Can you elaborate on why this matters? If it's in `/var/lib/microshift` then we would back it up, but we would also overwrite it when restoring. We could also remove it from the backup, if that is more reliable.

Reply:
Right, got lost in sequence of events.  If backup/restore manager reads and deletes that file before doing a backup/restore then backup won't have it, so it's good after all

Reply:
We cannot delete the file *before* an operation is complete because the system may crash in the middle of a data transaction. 
We can only flag operations complete after they are performed. In case of a failed operation, it will be redone on the next boot.

Reply:
That's good point

Reply:
Yes, backup should only contain data, not its status/flags metadata. We may want to review the directory structure to enable this as it allows for better extendability in the future, something like:
```
/var/lib/microshift
/var/lib/microshift.backup
   data.<AdeployID>
   data.<BdeployID>
   next.operation
   ...
```